### ISSUE TITLE
Probably buggy strategy ForwardToSelf / Restart

### ISSUE THREAD

**Issue by StefanBertels** [louthy/echo-process#37]
After doing some more tests I think there is something buggy with retry strategy:

```c#
 static void Main(string[] args)
        {
            ProcessConfig.initialise();

            spawn<Unit, Unit>("main", Setup: () =>
                {
                    spawn<string>("buggy", Inbox: _ =>
                    {
                        Console.WriteLine(DateTime.UtcNow + ": " + _);
                        throw new Exception(_);
                    });
                    return unit;
                },
                Inbox: (_, __) => _,
                
                Strategy: OneForOne(
                    Retries(Count: 5),
                    Backoff(Min: 2 * seconds, Max: 100 * seconds, Step: 0 * minute),
                    Always(Directive.Restart), // alternative: Resume
                    Redirect(MessageDirective.ForwardToSelf)) // alternative: StayInQueue
            );

            Console.WriteLine("actor runs");
            Task.Delay(2 * seconds).ConfigureAwait(false).GetAwaiter().GetResult();
            tell("/root/user/main/buggy", "hello");

            Console.WriteLine("ready for shutdown.");
            Console.ReadKey();
            Process.shutdownAll();
        }
```

Current guess: there might be an ordering problem in `Actor.cs` in `RunStrategy` at:

```c#
decision.Affects.Iter(p => pause(p));
                            safedelay(
                                () => RunProcessDirective(pid, sender, ex, message, decision),
                                decision.Pause
                            );
                            return InboxDirective.Pause | RunMessageDirective(pid, sender, decision, ex, message);
```

At least this should pass the faulty message to the inbox _before_ restarting the actor (due to Pause > 0) -- which might not be desired: first restart, i.e. repair, actor and _after_that_ re-process msg.

Anyway the code above does not what I expected regardless this order. I got two outputs directly one ofter the other and no more retries laster:

```
actor runs
ready for shutdown.
13.08.2018 09:38:15: hello
13.08.2018 09:38:15: hello
```

**Comment by StefanBertels** [comment]
Ok, I think the first bug ist that 

```
    public class PausableBlockingQueue<A> : IDisposable
    {
        readonly EventWaitHandle pauseWait = new AutoResetEvent(false);
```

should be initialised with true (so the first Pause will not be ignored): 

```
    public class PausableBlockingQueue<A> : IDisposable
    {
        readonly EventWaitHandle pauseWait = new AutoResetEvent(true);
```


There's still something going wrong because my actor does not call inboxFn again. I tried to follow the workflow and I think the problem is that there are calls to Pause/Unpause (IsPause = true/false) in Strategy code and other calls via InboxDirective.Pause in PausableBlockingQueue. There might be a problem with order of these calls as they run concurrently (via safedelay).

**Comment by StefanBertels** [comment]
Main problem / bugs fixed in PR #38.

Remaining question here: Is it by design that Restart+ForwardToSelf will cause "ForwardToSelf" first, then calling Restart later and when Restarting the unpause will process the re-sent messages in old actor (old state)?

**Comment by StefanBertels** [comment]
Additional note:

Restart / restart looks a bit inconsistent (see RunProcessDirective): siblings will get `restart(cpid)` which will result in Restart System message which will (in SystemMessageInbox) 1. unpause and 2. call Restart().

```
 switch (msg.Tag)
                {
                    case Message.TagSpec.Restart:
                        if (inbox.IsPaused)
                        {
                            inbox.Unpause();
                        }
                        actor.Restart();
                        break;
```

Question here: Why is Unpause() called here at all (and why before Restart)?

In RunProcessDirective there is a call to Actor.Restart()  (for the pid) which does no Unpause.

Naive view: if I call restart then Actors should not Unpause at all and one should call Unpause explicitely?

**Comment by louthy** [comment]
@StefanBertels Thanks for looking into the issues you've found.  I'll do my best to get to them as soon as I can.  I'm swamped with work at the moment.

**Comment by louthy** [comment]
@StefanBertels Thanks for your recent PRs.  They're all released now in `v2.0.49`
